feat: Add Supabase Integration

onurtemizkan opened the pull request:
Ref: https://github.com/getsentry/sentry-javascript/issues/15436

Summary: 

- Ports https://github.com/supabase-community/sentry-integration-js into `@sentry/core`
- Adds support for `auth` and `auth.admin` operations
- Adds browser integration tests
- Adds E2E tests running on NextJS

smeubank commented:
lookin good @onurtemizkan! part of getting this in will be updating the docs, and we can do some changelog post and updates to landing page, but i can help with all that it's simple

onurtemizkan commented:
@smeubank thanks, it would be great! We can then merge in queues support separately.

s1gr1d commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,488 @@
+  AUTH_ADMIN_OPERATIONS_TO_INSTRUMENT.forEach((operation: AuthAdminOperationName) => {
+    const authAdminOperation = auth.admin[operation];
+    if (typeof authAdminOperation === 'function') {
+      auth.admin[operation] = instrumentAuthOperation(authAdminOperation);
```
Doesn't this need the `isAdmin` as a second parameter if it is an admin operation? 🤔 

Maybe also add a test for the admin behavior.

s1gr1d commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,488 @@
+
+function instrumentSupabaseAuthClient(supabaseClientInstance: SupabaseClientInstance): void {
+  const auth = supabaseClientInstance.auth;
+
```
Do we need to check for `instrumented.has()` here as well?

s1gr1d commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,488 @@
+type AuthOperationFn = (...args: unknown[]) => Promise<unknown>;
+type AuthOperationName = (typeof AUTH_OPERATIONS_TO_INSTRUMENT)[number];
+type AuthAdminOperationName = (typeof AUTH_ADMIN_OPERATIONS_TO_INSTRUMENT)[number];
+type PostgrestQueryOperationName = (typeof AVAILABLE_OPERATIONS)[number];
```
Just to be sure: This is [PostgREST](https://postgrest.org/) and not [PostgreSQL](https://www.postgresql.org/)?

Super Nit: Call the variable `PostgREST....` so it's easier to read and understand. As it was lower-case, I read PostgreSQL at first sight.

Also valid for all other variable names in the file. But ignore this comment if you think this is not necessary.

smeubank replied:
I believe yes @s1gr1d. Some more context here

https://github.com/supabase-community/sentry-integration-js/issues/5#issuecomment-1959415384

s1gr1d commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,488 @@
+        const table = pathParts.length > 0 ? pathParts[pathParts.length - 1] : '';
+        const description = `from(${table})`;
+
+        const query: string[] = [];
```
As it's an array, I would suggest adding `items` to the name. "Query" kinda suggests it's a string.
```suggestion
        const queryItems: string[] = [];
```

s1gr1d commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,488 @@
+  }
+
+  // We need to wrap _all_ operations despite them sharing the same `PostgrestFilterBuilder`
+  // constructor, as we don't know which method will be called first, an we don't want to miss any calls.
```
```suggestion
  // constructor, as we don't know which method will be called first, and we don't want to miss any calls.
```

s1gr1d commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,488 @@
+  };
+}) satisfies IntegrationFn;
+
+export const supabaseIntegration = defineIntegration((supabaseClient: unknown) => {
```
Should the `supabaseIntegration` be added by default? If yes, then you should add it to `getDefaultIntegrations`.

Edit: Ah, as the `supabase` client needs to be passed, it actually cannot be default 🤔

s1gr1d commented on dev-packages/e2e-tests/test-applications/supabase-nextjs/README.md:
It's fine to delete this file if it's empty

s1gr1d commented on dev-packages/e2e-tests/test-applications/supabase-nextjs/lib/initSupabaseAdmin.ts:
```diff
@@ -0,0 +1,19 @@
+// These are the default development keys for a local Supabase instance
+const NEXT_PUBLIC_SUPABASE_URL = 'http://localhost:54321';
+const SUPABASE_SERVICE_ROLE_KEY =
+  'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImV4cCI6MTk4MzgxMjk5Nn0.EGIM96RAZx35lJzdJsyH-qQwv8Hdp7fsn3W0YpN81IU';
```
Is it okay to leave this here? Or should we rather add this to the environment?

Same for `NEXT_PUBLIC_SUPABASE_ANON_KEY`

onurtemizkan replied:
These are default local supabase server keys. I think it's ok to leave them here.

lforst commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,493 @@
+          queryItems.push(translateFiltersIntoMethods(key, value));
+        }
+
+        const body: Record<string, unknown> = {};
```
```suggestion
        const body: Record<string, unknown> = Object.create(null);
```

to prevent prototype pollution.

lforst reviewed (commented):
A first pass - but it's already looking great!

smeubank commented:
@onurtemizkan do you have an example in team-sdks or something for this? I am just curious what this looks like in product, to also look at for the flutter integration later

> Adds support for auth and auth.admin operations

onurtemizkan commented:
@smeubank, sure, I'll create example events and post links here after making the reviewed changes today 👍

smeubank commented:
> @smeubank, sure, I'll create example events and post links here after making the reviewed changes today 👍

https://github.com/smeubank/workout-app-next/pull/1

@onurtemizkan i created a simple example app that i deployed on vercel, it's pointed at my private repo, I'll eventually add auth too. Added you as contributor so you can see the private project in case it is helpful. This PR is for the old integration which is a total pain to set up 😅 

for the new integration do we know already what the users has to do to get this integration working? Is it just in the init? is it auto?

Also wondering for already updating docs, for sentry and for supabase, cause we'll replace [this page ](https://supabase.com/docs/guides/telemetry/sentry-monitoring)eventually

onurtemizkan commented:
@smeubank - it's not auto, though it's easy to integrate.

Now after the review updates, it's currently set up by:

- If there are multiple `supabaseClient` instances per Sentry client (browser / node / edge?), or the `supabaseClient` is not available from where we init Sentry, we can just do,

`Sentry.instrumentSupabase({supabaseClient: supabaseClient})`

without using the integration at `Sentry.init`(s)

- If not, we can init Sentry like:

```typescript
const supabaseClient = createClient('...', 'test-key');

Sentry.init({
  dsn: 'https://public@dsn.ingest.sentry.io/1337',
  integrations: [
    Sentry.browserTracingIntegration(),
    Sentry.supabaseIntegration({
      supabaseClient
    })
  ],
  tracesSampleRate: 1.0,
});
```

I think, at this point we don't need to make this an integration. We can just expose `Sentry.instrumentSupabase` from relevant packages.

I'll push the final updates here and will open a PR there 👍

onurtemizkan commented:
@smeubank - sample traces with the current implementation:

- Auth (Admin):  [Link](https://sentry-sdks.sentry.io/traces/trace/e02c3a4db443fdf2cd391732b99eef81/?pageEnd&pageStart&project=5659328&source=traces&statsPeriod=1h&table=trace&timestamp=1744307516.429)
- Auth / DB: [Link](https://sentry-sdks.sentry.io/traces/trace/c76f9a21a64b16886fb80e1da17cf4d1/?pageEnd&pageStart&project=5659328&source=traces&statsPeriod=1h&table=trace&timestamp=1744307515.919)

smeubank commented:
@s1gr1d @onurtemizkan and @lforst for whomever writes/reviews

We should ensure that we are populating db-insights in product. Below the link to the span eligibilty section of docs, 
- span op looks correct mostly, and tbh I think those requirements should be expanded in product.
- span description with query, i think right now is not always parametrized, i don't know how much we can fix that in the SDK though. Or if we need ingest too
- db.system: should be easy fix we need to set `postgresql`

https://docs.sentry.io/product/insights/backend/queries/#span-eligibility

onurtemizkan commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,493 @@
+  ) => Promise<T>;
+}
+
+const instrumented = new Map();
```
I removed this and flagged instrumented objects instead, which is more reliable for browser/client and multiple clients usage.

onurtemizkan commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,493 @@
+          attributes,
+        });
+
+        return (Reflect.apply(target, thisArg, []) as Promise<SupabaseResponse>)
```
I tried this, and it breaks the code. I did not debug that much, and it's from the original implementation. The arguments are passed in the chained `then`  at the bottom.

onurtemizkan commented:
@lforst - I updated the PR addressing the comments. I have exported `instrumentSupabase` and ended up thinking that we may not even need an 'integration' for this use case. Most of the uses I have seen have multiple client instances and recreating instances looks like the norm. We can still export the integration for whenever it's needed, though.

lforst commented on packages/core/src/integrations/supabase.ts:
```diff
@@ -0,0 +1,516 @@
+  }
+}
+
+export const instrumentSupabase = (options: { supabaseClient: unknown }): void => {
```
I would almost vote that we make the signature `instrumentSupabaseClient(client: unkown)` so we can make it `instrumentSupabaseClient(client: unkown, options: Options = {})` in the future if we ever want to do options.

lforst reviewed (approved):
One last minor thing then 🚀

onurtemizkan commented:
@lforst, I'm updating the types that are failing on TS 3.8

lforst commented:
@onurtemizkan I also removed the exports from the browser bundles. I don't think we want that *yet*. Maybe if people come asking!

harrisrobin commented:
so how do we use it with nextjs? After upgrading, i do not see it exported as part of `@sentry/nextjs`

onurtemizkan commented:
@harrisrobin, it [should be exported](https://github.com/getsentry/sentry-javascript/blob/640b57f16a72d12295c17ed7fc07cc53d0a9d232/dev-packages/e2e-tests/test-applications/supabase-nextjs/lib/initSupabaseAnon.ts) starting from version `9.14.0`. Can you open a separate issue with a reproduction if the versions are correct and you still can't access it?

chollier commented:
does this support Supabase edge functions tracing/instrumentation?
